### DISCUSSION TITLE
Vulkan dependency detection when cross compiling (Yocto)

### DISCUSSION THREAD

**Discussion by jwinarske** [mesonbuild/meson#9673] [Q&A] [answered]
When detecting Vulkan dependency in cross compile case (Yocto) what is the proper pattern for deriving the absolute includedir?

I'm doing this:
```
vulkan_dep = dependency('vulkan')
vulkan_hpp = join_paths([
    vulkan_dep.get_pkgconfig_variable('includedir'),
    'vulkan',
    'vulkan.hpp'
    ])
```

How do I get sysroot to affect this, as I always end up with `/usr/include/vulkan/vulkan.hpp`.

**Comment by dcbaker** [comment]
Is the vulkan dependency correct in that case at all, or is it just getting the pkg-config variable that is broken? This could be a bug

**Reply by jwinarske** [reply]
I tried a number of scenarios:


1.
```
$ pkg-config --print-errors --define-variable=prefix=/opt --variable=includedir vulkan
/usr/include
```
Equivalent to this
```
vulkan_hpp = join_paths([
    vulkan_dep.get_pkgconfig_variable('includedir', define_variable: ['prefix', get_option('prefix')])
    'vulkan',
    'vulkan.hpp'
    ])
```
Setting variable in this manner has no impact.

2.
```
$ export STAGING_DIR_TARGET=/b/github-ci/_work/meta-flutter/rpi4-drm-honister-latest/raspberrypi4-64/tmp/work/cortexa72-poky-linux/vkmark/git-r0/recipe-sysroot
$ PKG_CONFIG_SYSROOT_DIR=$STAGING_DIR_TARGET pkg-config --print-errors --define-variable=prefix=/opt --variable=includedir vulkan
/b/github-ci/_work/meta-flutter/rpi4-drm-honister-latest/raspberrypi4-64/tmp/work/cortexa72-poky-linux/vkmark/git-r0/recipe-sysroot/usr/include
```

This behavior matches the docs, and indicates that meson is not passing `PKG_CONFIG*` variables to shell executing pkg-config.


3.  Then in Yocto meson cross file I added a `sys_root` with value in `properties` section of meson.cross.  This directly causes `PKG_CONFIG_SYSROOT_DIR` to be set based on the build log.  Only pkg-config can't use it.

So if meson were to pass `PKG_CONFIG*` environmental variables to the shell that runs pkg-config, it would work as expected.

If that were to happen then for near term  the work around for Yocto recipe could be:

```
do_configure:prepend() {
    export PKG_CONFIG_SYSROOT_DIR="${STAGING_DIR_TARGET}"
}
```

For long term fix would be to update meson.bbclass to write `sys_root = '${STAGING_DIR_TARGET}'` under the properties section of meson.cross.

**Reply by jwinarske** [reply]
Also the value returned from `vulkan_dep.get_pkgconfig_variable` is the literal value in the .pc file.

**Reply by jwinarske** [reply]
Useful references:
https://linux.die.net/man/1/pkg-config
https://autotools.io/pkgconfig/cross-compiling.html

**Reply by jwinarske** [reply]
I should also note when building Yocto recipe on machine with vulkan-loader installed (vulkan.pc) it succeeds, whereas when I build it in a docker container that does not have vulkan-loader installed, it fails.  So the pkg-config invocation appears to not using target sysroot.  It just happens that both system and target vulkan.pc have identical includedir values.  If they were different the build would certainly fail.

**Reply by dcbaker** [reply]
What does your vulkan.pc look like? When I look at mine, the pc file is actually broken. It should have something like:
```txt
prefix=/usr
includedir=${prefix}/include
libdir=${prefix}/lib
```

but mine has:
```txt
prefix=/usr
includedir=/usr/include
libdir=/usr/lib
```
So even though we set the variables they don't work since the pkg-config file itself is just broken. If yours is the same then this a bug with the vulkan loader and we need to move that there.

**Reply by jwinarske** [reply]
It's the same... 

This is clearly wrong:
https://github.com/KhronosGroup/Vulkan-Loader/blob/master/loader/vulkan.pc.in

I'll submit a PR.

https://github.com/KhronosGroup/Vulkan-Loader/issues/490

**Reply by dcbaker** [reply]
Please mention me on that, I'm a khronos member so my review there might speed things up :)

**Reply by jwinarske** [reply]
https://github.com/KhronosGroup/Vulkan-Loader/pull/756

**Reply by jwinarske** [reply]
merged

**Reply by dcbaker** [reply]
I went ahead and marked this as answered

**Comment by jwinarske** [comment]
While now I have a correct vulkan.pc meson is still ignoring PKG_CONFIG_SYSROOT_DIR and returning the same value as before.  If PKG_CONFIG_SYSROOT_DIR is set pkg-config should honor it, which indicates it's not available at runtime.

```
vulkan_hpp = join_paths([
    vulkan_dep.get_pkgconfig_variable('includedir'),
    'vulkan',
    'vulkan.hpp'
    ])
```

Re-opened issue: https://github.com/mesonbuild/meson/issues/9674

**Comment by jwinarske** [comment, marked as answer]
This is the functional pattern

1. Correctly implemented package config file. @dcbaker thank you.

2. Meson
```
vulkan_dep = dependency('vulkan')
vulkan_hpp = join_paths([
    vulkan_dep.get_pkgconfig_variable('includedir', define_variable: ['prefix', get_option('prefix')]),
    'vulkan',
    'vulkan.hpp'
    ])
```

3. Yocto Recipe
```
EXTRA_OEMESON += "--prefix ${STAGING_DIR_TARGET}/usr"
```

This project now correctly builds on desktop and Yocto:  https://github.com/jwinarske/vkmark/tree/jw/vulkan_dep
